edalize/templates/apicula/apicula-makefile.j2

```diff
@@ -0,0 +1,21 @@
+#Auto generated by Edalize
+
+TARGET   := {{ name }}
+CST_FILE := {{ cst_file }}
+NEXTPNR_OPTIONS     := {{ nextpnr_options|join(' ') }}
+
+all: $(TARGET).fs
+
+%.json: %.mk
+	$(MAKE) -f $<
+%.pack: %.json
+	nextpnr-gowin -l next.log --device $(DEVICE)  $(NEXTPNR_OPTIONS) --cst $(CST_FILE) --json $? --write $@
+%.fs: %.pack
+	gowin_pack -d GW1N-1 -o  $@ $^
```

Comment:
One problem I see now is that the model GW1N-1 is hardcoded in the makefile for packing.  Maybe it will be necessary to add an option in the .core file for packing like nextpnr_options : [GW1N-LV1QN48C6/I5] for example.
I will wait for your feedback before I start to work on this.

Reply:
Good catch. I didn't see that in my first review. Yes, we need to remove this hardcoded value, but I'm a bit confused about how to do this. Most of the FPGA toolchain backends only require a single `part` option to decide which device to target. The yosys+nextpnr variants however are a bit different. Because yosys doesn't seem to care at all (and arachne-pnr/nextpnr used slighlty different names for parts) we have previously just defined the part in nextpnr_options. I was looking at how the other pack commands worked and it seems that gowin_pack is the only one that explicitly needs a part name here. So in that case we need to inform both nextpnr and gowin_pack about the part name, and I'm not sure yet if the are using the same naming schemes. I really want to avoid having some lookup table in the backend. Perhaps @pepijndevos could shed some light on this.

Reply:
Yea so Yosys doesn't care at all. Nextpnr takes the whole part number because it cares about timing and such. The packer only needs the family because they have different padding values. So the part number is `GW1N{variant}-(LV|UV){luts}{package}{speed}` And the packer only needs the `GW1N-{luts}` part. (eventually GW2A maybe). If this is problematic I'm open to a PR making the packer take the whole part number, or maybe even encode the part number in the nextpnr output somehow.

https://github.com/YosysHQ/nextpnr/blob/master/gowin/main.cc#L57

Reply:
I will continue to work on the gowin backend when there will be a solution for setting the part number for gowin_pack. I'm afraid that I have not enough skills right now to make a PR for the packer problem.

Reply:
Yes, I think that's the best thing to do right now. It's a small thing but better to get right from the start.
@daveshah1, @clairexen how does ecppack and icepack handle this?

Reply:
It's actually a pretty trivial fix. Might do it this weekend. Basically just copy-paste the linked regex and add a new `--part` option that uses the regex to set the device to `GW1N-{luts}`. https://github.com/YosysHQ/apicula/blob/master/apycula/gowin_pack.py#L126

Reply:
Ok, so if I understand correctly, with such a fix applied, we could pass the same part number to nextpnr and gowin_pack? If so, that would be great.

For reference, I just looked at icepack ecppack and they both seem to look at the `.device` line in the input file. Could that be an option for apicula too?

Reply:
Done https://github.com/YosysHQ/apicula/commit/7624fdcd4fbebe880296aba4d54818e243ddb0a4

icepack and ecppack both use a custom text format for writing nextpnr output, while Apicula uses the built-in json output. This does not contain the part number. Writing a custom fasm exporter is not high on my priority list...

Reply:
Awesome! That's all we need to proceed.

For my understanding, is the built-in json format fasm, or are those two different formats? And would the long-term solution be to add a device field tl this format?

Reply:
No fasm is a completely different thing: https://github.com/SymbiFlow/fasm
So moving from JSON to FASM might be a veeeeery long term plan, but it only really makes sense if there are multiple competing PnR implementations. JSON is very Nextpnr-specific. VTR uses fams, but I don't have any plans to add a VTR backend. So JSON is fine with me.

Reply:
Got it. Thanks